### PR TITLE
use vectorized io when writing dirty pages

### PR DESCRIPTION
**Pull request by tjungblu**
this is still a bit experimental, thus drafting it - but that should reduce the amount of write syscalls by a lot and resolve a very old TODO.

Will benchmark this more in-depth tomorrow, but if somebody finds some time earlier I would appreciate some other tests as well!

### CODE REVIEW

File: bolt_windows.go [hunk 1]
```diff
@@ -10,6 +10,10 @@ import (
 	"golang.org/x/sys/windows"
 )
 
+func init() {
+	vectorizedWritesAvailable = false
```
Comment (tjungblu) [inline bolt_windows.go:14]: also not sure whether this is the best way to handle the availability of this feature

**Comment by tjungblu** [comment]
also coded up an alternative with existing syscall in golang/x/sys that works with `[][]byte`:
https://github.com/etcd-io/bbolt/compare/master...tjungblu:bbolt:vectorizedio2

This one, interestingly enough, has some issues in the manydb tests. It leads me to believe that gogc is re-arranging the pages and that makes it write some weird memory into the file. Consistency checks here fail too.

**Comment by ahrtr** [comment]
Please ping me when you done.  

Also reminder, small PR (e.g. < 100 lines of change) is preferred.

**Comment by tjungblu** [comment]
@ahrtr I tried to simplify a bit more, just over 100 lines added though. I still have to benchmark this further... Last time in September I remember it was 10% slower.

**Comment by tjungblu** [comment]
just trying to get my benchmark setup and running again, I get a minor difference in tail laltency:

```
using `etcdctl check perf`:

3.5.6 performance:
[tjungblu ~]$ etcdctl --endpoints=hserver:31567 check perf
 59 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooom  !  98.33%PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.110571s
PASS: Stddev is 0.003879s

3.5.6 performance with this bbolt PR:
[tjungblu ~]$ etcdctl --endpoints=hserver:32292 check perf
 59 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooom  !  98.33%PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.117600s
PASS: Stddev is 0.003616s

```

I'm running a couple of strace sessions now to see how many syscalls this actually saves. Ideally this should reduce tail latency on very big consecutive writes.

**Comment by xiang90** [comment]
I would not expect a huge performance boost for the light write workload. But it should not be 10% slower, either :).

**Comment by tjungblu** [comment]
Let me triple check the docker images again, with strace I couldn't really see the new syscall being used. It's still really difficult to get the build to get the right bbolt version, even with go mods and replaces :/

**Comment by tjungblu** [comment]
Verified that the throughput and latency is not affected by this PR at all, WAL seems to be the dominating factor.

Running some tests today on Intel Cascade Lake 4 core with 100g of SSD persistent disk on GCP, another same-spec'd machine in the same VPC runs the check commands using etcdctl.

image

Above is a run of three operations each, spaced by a minute or two:
> check datascale --load=l
> check perf --load=xl
> check perf --load=xl

Left side starting with this PR on top of 3.5.6, the right hand side after 12.35pm is vanilla 3.5.6. So not that much of an improvement I'd expected to see. 

One thing I would still like to test is the defrag performance, this should be doing pretty well as it re-arranges the pages again into a pattern that would benefit pwritev. Having said that, it's probably not worth optimizing this case specifically.

**Comment by tjungblu** [comment]
rebasing over 3.6, the pattern is similar :/

regarding defrag, with 3.6 (current HEAD) and a db size of 5.4gb:
> {"level":"info","ts":"2022-11-30T14:12:44.978Z","caller":"backend/backend.go:547","msg":"finished defragmenting directory","path":"default.etcd/member/snap/db","current-db-size-bytes-diff":-24576,"current-db-size-bytes":5432389632,"current-db-size":"5.4 GB","current-db-size-in-use-bytes-diff":-24576,"current-db-size-in-use-bytes":5432369152,"current-db-size-in-use":"5.4 GB","took":"1m3.791887536s"}

with this bbolt patch on the same database:
> {"level":"info","ts":"2022-11-30T14:15:31.033Z","caller":"backend/backend.go:547","msg":"finished defragmenting directory","path":"default.etcd/member/snap/db","current-db-size-bytes-diff":-24576,"current-db-size-bytes":5432389632,"current-db-size":"5.4 GB","current-db-size-in-use-bytes-diff":-24576,"current-db-size-in-use-bytes":5432369152,"current-db-size-in-use":"5.4 GB","took":"30.260298509s"}

So that seems to be the clear winner operation here ;-)

**Comment by xiang90** [comment]
@tjungblu 

Ah... I think you probably should just run a benchmark for the boltdb rather than etcd? I remember there is a benchmark subcommand for boltdb in bolt ctl or something :P. Then the WAL stuff will not affect your result.

**Comment by xiang90** [comment]
@tjungblu 

Here https://github.com/etcd-io/bbolt/blob/master/cmd/bbolt/main.go#L146

**Comment by tjungblu** [comment]
@xiang90 beautiful, thanks for digging this up. I'll benchmark this in depth tomorrow using bbolt bench and post the result.

File: bolt_unix.go [hunk 2]
```diff
@@ -12,6 +12,59 @@ import (
...
+	// TODO: read from this from sysconf(_SC_IOV_MAX)?
+	// linux and darwin default is 1024
+	const maxVec = 1024
+
+	lastPid := pages[0].id - 1
+	begin := 0
+	var curVecs []syscall.Iovec
+	for i := 0; i < len(pages); i++ {
+		p := pages[i]
+		if p.id != (lastPid+1) || len(curVecs) >= maxVec {
```
Comment (ahrtr) [inline bolt_unix.go:47]: This seems not correct. Shouldn't it be something like below? 
```
p.id != lastPid + last_overflow + 1
```

File: bolt_unix.go [hunk 3]
```diff
@@ -12,6 +12,59 @@ import (
...
+// The logic identifies runs of consecutive pages that are written with vectorized io using the pwritev syscall.
+func pwritev(db *DB, pages []*page) error {
+	offsets, iovecs := pagesToIovec(db, pages)
+
+	var (
+		e syscall.Errno
+	)
+	for i := 0; i < len(offsets); i++ {
+		_, _, e = syscall.Syscall6(syscall.SYS_PWRITEV, db.file.Fd(), uintptr(unsafe.Pointer(&iovecs[i][0])),
+			uintptr(len(iovecs[i])), uintptr(offsets[i]), uintptr(offsets[i]>>0x8), 0)
```
Comment (ahrtr) [inline bolt_unix.go:25]: The syscall pacakge is deprecated and locked down. Please consider using `golang.org/x/sys` instead. I had a quick look at your another solution [vectorizedio2](https://github.com/etcd-io/bbolt/compare/master...tjungblu:bbolt:vectorizedio2), it seems that you did not cast `iovecs` to [][]byte correctly (about line 556 in tx.go).

Comment (tjungblu) [reply bolt_unix.go:25]: yeah, let me try again with the other solution. I'm afraid this will be drastically slower due to the extra allocation however.

File: bolt_unix.go (hunk 2 above)
Comment (tjungblu) [reply bolt_unix.go:47]: interesting, let me see how the overflow works. I was under the impression that the page id is just an increasing identifier for the page.

Comment (tjungblu) [reply bolt_unix.go:51]: fixed, very curious that no e2e test case covered this. I'm also not even sure whether the overflow is in use.

**Comment by tjungblu** [comment]
with bench the difference is clearer:

```
this PR:
[tjungblu ~/git/bbolt]$ go run cmd/bbolt/main.go bench -count 100000
# Write	173.741269ms	(1.737µs/op)	(575705 op/sec)
# Read	1.001078252s	(14ns/op)	(71428571 op/sec)


master:
[tjungblu ~/git/bbolt]$ go run cmd/bbolt/main.go bench -count 100000
# Write	184.005891ms	(1.84µs/op)	(543478 op/sec)
# Read	1.000812791s	(14ns/op)	(71428571 op/sec)

```

File: bolt_unix.go (hunk 3 above)
Comment (tjungblu) [reply bolt_unix.go:29]: yep, it's significantly slower:

```
with x/sys:
[tjungblu ~/git/bbolt]$ go run cmd/bbolt/main.go bench -count 100000
# Write	199.840821ms	(1.998µs/op)	(500500 op/sec)
# Read	1.000347769s	(17ns/op)	(58823529 op/sec)


this PR:
[tjungblu ~/git/bbolt]$ go run cmd/bbolt/main.go bench -count 100000
# Write	173.741269ms	(1.737µs/op)	(575705 op/sec)
# Read	1.001078252s	(14ns/op)	(71428571 op/sec)


master:
[tjungblu ~/git/bbolt]$ go run cmd/bbolt/main.go bench -count 100000
# Write	184.005891ms	(1.84µs/op)	(543478 op/sec)
# Read	1.000812791s	(14ns/op)	(71428571 op/sec)

```

Comment (tjungblu) [reply bolt_unix.go:29]: pushed the code into another branch https://github.com/etcd-io/bbolt/compare/master...tjungblu:bbolt:vec2

**Comment by tjungblu** [comment]
seems the vectorized write in the bbolt benchmark is also not a bottleneck:

image

image

Let's stash this PR then for the time being, seems we can get more performance by optimizing the allocation. Thanks for your help everyone!

**Comment by xiang90** [comment]
From the CPU perspective, yes, most of the computing is probably spent on `allocate`. We made some improvements on that, but there must be tons that can be further improved... On the I/O path, I feel there are probably even more inefficiencies.

**Comment by ptabor** [comment]
Likely I'm missing something obvious: 
  - we have the bbolt disk file mmaped... 
  - we only use the mmap area to make reads (db.go:933)
  - we don't we use it perform writes (i.e. copy onto the memory) and call [msync](https://linux.die.net/man/2/msync)
Is there a good reason ?

Even further: node.write() should directly write to the db.data (mmapped buffer) instead of working on intermediate buffer.

**Comment by xiang90** [comment]
@ptabor 

This behavior is from LMDB. And there are reasons for it. This [link](https://news.ycombinator.com/item?id=19127011) provides some of the benefits of handling writes explicitly.

**Comment by ptabor** [comment]
Thank you @xiang90. I've read the article, but it mostly says about concurrent calls of fsync and responsibility of retries. 

My mental model is following (assuming the goal is to get highest possible performance [and we are not in position to complicate bbolt's reliability for performance] ): 

  1. The bbolt writes/syncs are single-threaded, so there is an exclusive lock under which we need to make sure transaction is committed, it is fync or msync the file. 
  2. We need to precisely control write of the meta pages. It's meta page can be only written if we make sure all other modifications to the tree are persisted. So writes to meta pages should be handled explicitly. 
  3. All other modifications than meta page should be not 'referencable' so the order of writes does not matter. 
  4. work on mmaped pages would allow to: 
      - avoid allocation of temporary memory for pages and 'zeroing' them
      - avoid copy of the temporary content into fwrite() memory. 
      - the additional risk & benefit is that mmaped content can start being written to disk as soon as we (partially) modifying the pages -> without delaying it till the commit happens. This is optional mode (alternatively we can do copy-on-write), but it's sense depends on the ratio of rolled back transactions.  For etcd there are nearly no rolled back transactions. 

Again - I'm not saying we should do it... Just brainstorming what's the "speed of light" version of the memory copying optimization.

**Comment by ptabor** [comment]
@tjungblu  
From the chart https://github.com/etcd-io/bbolt/pull/339#issuecomment-1333904758, the time seems to be dominated by mmap. But the mmap should be an infrequent operation for databases that are 'mature', as it happens only when the file needs to grow.  Wouldn't it be better to benchmark an already upsized file, e.g. run with high `InitialMmapSize` option ?

**Comment by xiang90** [comment]
@ptabor 

Right. I actually referred to [hyc_symas](https://news.ycombinator.com/user?id=hyc_symas)'s (he is the author of LMDB) comment in that HN thread, not just the article :P. 

But I think you are probably right. Only the meta-pages must be handled explicitly (split from the normal pages mmap sync). Other pages should not matter much, assuming msync will return the error explicitly for any failing physical writes. I do not see a reason why it cannot be done. But we may want to understand the perf benefits as well :)
